File: src/main/java/com/baulsupp/oksocial/completion/BaseUrlCompleter.java
package com.baulsupp.oksocial.completion;

import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;
import okhttp3.HttpUrl;

import static java.util.concurrent.CompletableFuture.completedFuture;

public class BaseUrlCompleter extends HostUrlCompleter {
  private final UrlList urlList;
  private CompletionMappings mappings = new CompletionMappings();

  public BaseUrlCompleter(UrlList urlList, Collection<String> hosts) {
    super(hosts);
    this.urlList = urlList;
  }

  @Override public CompletableFuture<UrlList> siteUrls(HttpUrl url) throws IOException {
    CompletableFuture<UrlList> future = completedFuture(urlList);

    return future.thenCompose(r -> mappings.replaceVariables(r));
  }

  public void withVariable(String name, List<String> values) {
    mappings.withVariable(name, values);
  }

  public void withVariable(String name, Supplier<CompletableFuture<List<String>>> values) {
    mappings.withVariable(name, values);
  }
}

File: src/main/java/com/baulsupp/oksocial/completion/BaseUrlCompleter.kt
package com.baulsupp.oksocial.completion

import java.io.IOException
import java.util.concurrent.CompletableFuture
import java.util.function.Supplier
import okhttp3.HttpUrl

import java.util.concurrent.CompletableFuture.completedFuture

class BaseUrlCompleter(private val urlList: UrlList, hosts: Collection<String>) : HostUrlCompleter(hosts) {
    private val mappings = CompletionMappings()

    @Throws(IOException::class)
    override fun siteUrls(url: HttpUrl): CompletableFuture<UrlList> {
        val future = completedFuture(urlList)

        return future.thenCompose { r -> mappings.replaceVariables(r) }
    }

    fun withVariable(name: String, values: List<String>) {
        mappings.withVariable(name, values)
    }

    fun withVariable(name: String, values: Supplier<CompletableFuture<List<String>>>) {
        mappings.withVariable(name, values)
    }
}
